src/Composer/Util/Git.php

```diff
@@ -300,11 +301,25 @@ public function syncMirror($url, $dir)
         // clean up directory and do a fresh clone into it
         $this->filesystem->removeDirectory($dir);
 
-        $commandCallable = function ($url) use ($dir) {
-            return sprintf('git clone --mirror -- %s %s', ProcessExecutor::escape($url), ProcessExecutor::escape($dir));
-        };
+        if ($shallowClone) {
+            $commandCallable = function ($url) use ($dir) {
+                return sprintf('git clone --depth=1 --mirror -- %s %s && git remote set-branches origin "*"', ProcessExecutor::escape($url), ProcessExecutor::escape($dir));
```

Comment:
Sorry but I'm not familiar with this here, why is it doing `git remote set-branches origin "*"` in addition to the clone? And why the additional fetch --depth=1 below? Is that to force fetching one commit of every branch?

In any case, this needs cross-platform escaping:

```suggestion
                return sprintf('git clone --depth=1 --mirror -- %s %s && git remote set-branches origin -- %s', ProcessExecutor::escape($url), ProcessExecutor::escape($dir), ProcessExecutor::escape('*'));
```

Reply:
`git remote set-branches origin "*"`  is to fetch all the remote branches, without it I only got the default branch and the additional fetch is to fetch one commit for every branch. I'm also not super with how this works but found this https://stackoverflow.com/questions/23708231/git-shallow-clone-clone-depth-misses-remote-branches/27393574#27393574 while searching for how to fetch all branches

Reply:
OK thanks for confirming, looks good to me then, thanks!